pkg/daemon/update.go

```diff
@@ -306,53 +307,36 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) (retErr err
 		}
 	}()
 
-	if err := dn.updateFIPS(oldConfig, newConfig); err != nil {
+	if err := dn.checkFIPS(oldConfig, newConfig); err != nil {
 		return err
 	}
-	defer func() {
-		if retErr != nil {
-			if err := dn.updateFIPS(newConfig, oldConfig); err != nil {
-				retErr = errors.Wrapf(retErr, "error rolling back FIPS %v", err)
-				return
-			}
-		}
-	}()
 
 	return dn.updateOSAndReboot(newConfig)
 }
 
-func (dn *Daemon) updateFIPS(current, desired *mcfgv1.MachineConfig) error {
-	if current.Spec.FIPS == desired.Spec.FIPS {
-		return nil
-	}
-	if dn.OperatingSystem != machineConfigDaemonOSRHCOS {
-		return errors.New("Updating FIPS on non-RHCOS nodes is not supported")
-	}
+func (dn *Daemon) checkFIPS(current, desired *mcfgv1.MachineConfig) error {
 	// Our new thought around this is that really FIPS should be a "day 1"
-	// operation, and we don't want to make it really easy to undo.
+	// operation, and we don't want to make it editable after the fact.
 	// See also https://github.com/openshift/installer/pull/2594
 	// Anyone who wants to force this can change the MC flag, then
 	// `oc debug node` and run the disable command by hand, then reboot.
-	if current.Spec.FIPS && !desired.Spec.FIPS {
-		return errors.New("Refusing to undo FIPS mode")
-	}
-	// At this point, we must be trying to enable FIPS, since
-	// current != desired && desired per conditionals above
-	if _, err := os.Stat(fipsCommand); err != nil {
-		if os.IsNotExist(err) {
-			return errors.New("Cannot enable FIPS after firstboot")
-		}
-		return errors.Wrapf(err, "Checking FIPS")
-	}
+	// If we detect that FIPS has been changed, we reject the update.
 
-	arg := "enable"
-	if !desired.Spec.FIPS {
-		arg = "disable"
+	content, err := ioutil.ReadFile(fipsFile)
+	if err != nil {
+		return errors.Wrapf(err, "Error reading FIPS file at %s: %s", fipsFile, string(content))
 	}
-	cmd := exec.Command(fipsCommand, arg)
-	dn.logSystem("Running %s %s", fipsCommand, arg)
-	if out, err := cmd.CombinedOutput(); err != nil {
-		return errors.Wrapf(err, "%s FIPS: %s", arg, string(out))
+	nodeFIPS, err := strconv.ParseBool(string(content))
+	if err != nil {
+		return errors.Wrapf(err, "Error parsing FIPS file at %s", fipsFile,)
+	}
+	if desired.Spec.FIPS == nodeFIPS {
+		// Check if FIPS on the system is at the desired setting
+		return nil
+	}
+
+	if current.Spec.FIPS != desired.Spec.FIPS {
```

Comment:
Hmm, should we move this check higher up?

Reply:
Hm so we ran into this error initially because we detected that the current `current.Spec.FIPS` != `desired.Spec.FIPS` right? Because it was nil?

Reply:
As in, maybe I should change this just to return an error if `!desired.Spec.FIPS == nodeFIPS`

Reply:
And flat out ignore what `current.Spec.FIPS` is

Reply:
but then the Spec would not be reflective of the state of the system, perhaps I should update that as well just above?

Reply:
I changed it to update current.Spec.FIPS so we don't get stuck in a reboot loop. This should definitely be improved of course but I think this can get us past the errors (not tested yet)